services/web/server/src/simcore_service_webserver/projects/_ports.py

```diff
@@ -0,0 +1,157 @@
+from dataclasses import dataclass
+from typing import Any, Iterator, Literal, Optional
+
+from models_library.function_services_catalog.api import catalog
+from models_library.projects_nodes import Node, NodeID
+from models_library.projects_nodes_io import PortLink
+from models_library.utils.json_schema import (
+    JsonSchemaValidationError,
+    jsonschema_validate_data,
+)
+from models_library.utils.services_io import get_service_io_json_schema
+from pydantic import ValidationError
+
+
+@dataclass(frozen=True)
+class _ProjectPort:
+    kind: Literal["input", "output"]
+    node_id: NodeID
+    io_key: str
+    node: Node
+
+    @property
+    def key(self):
+        return f"{self.node_id}.{self.io_key}"
+
+    def get_schema(self) -> Optional[dict[str, Any]]:
+        node_meta = catalog.get_metadata(self.node.key, self.node.version)
+        if self.kind == "input" and node_meta.outputs:
+            if input_meta := node_meta.outputs[self.io_key]:
+                return get_service_io_json_schema(input_meta)
+        elif self.kind == "output" and node_meta.inputs:
+            if output_meta := node_meta.inputs[self.io_key]:
+                return get_service_io_json_schema(output_meta)
+        return None
+
+
+def _iter_project_ports(
+    workbench: dict[NodeID, Node],
+    filter_kind: Optional[Literal["input", "output"]] = None,
+) -> Iterator[_ProjectPort]:
+    """Iterates the nodes in a workbench that define the input/output ports of a project
+
+    - A node in general has inputs and outputs:
+      - Inputs can have values or references.
+      - Outputs have only values
+
+    - A project defines its ports with special nodes:
+        - An input node: has no inputs and stores its value as an output
+        - An output node: has not outputs and stores a reference to an output in his input
+    """
+
+    for node_id, node in workbench.items():
+        is_inputs = filter_kind is None or filter_kind == "input"
+        is_outputs = filter_kind is None or filter_kind == "output"
```

Comment:
Pleas make the naming all singular or plural. Either `is_input`  or `are_inputs`, same with `is_outputs`